File: PowerUp/PowerUp.ElectronUI/electron/src/app/teamEditor/playerDetailsResponse.ts
import { EntitySourceType } from "../shared/entitySourceType";
import { GeneratorWarningDto } from "../shared/generatorWarning";
import { Position } from "../shared/positionCode";
import { PlayerDetails } from "./playerRoleState";

export interface PlayerDetailsResponse {
  sourceType: EntitySourceType;
  canEdit: boolean;
  playerId: number;
  fullName: string;
  savedName: string;
  generatedPlayer_Warnings: GeneratorWarningDto[];
  generatedPlayer_IsUnedited: boolean;
  position: Position;
  positionAbbreviation: string;
  overall: number;
  pitcherType: string;
  batsAndThrows: string;
  throwingArm: string;
  topSpeed: string;
  control: string;
  stamina: string;
}

export function toPlayerDetails(detailsResponse: PlayerDetailsResponse): PlayerDetails {
  return {
    sourceType: detailsResponse.sourceType,
    canEdit: detailsResponse.canEdit,
    playerId: detailsResponse.playerId,
    fullName: detailsResponse.fullName,
    savedName: detailsResponse.savedName,
    generatedPlayer_Warnings: detailsResponse.generatedPlayer_Warnings,
    generatedPlayer_IsUnedited: detailsResponse.generatedPlayer_IsUnedited,
    position: detailsResponse.position,
    positionAbbreviation: detailsResponse.positionAbbreviation,
    overall: detailsResponse.overall,
    batsAndThrows: detailsResponse.batsAndThrows,
    pitcherType: detailsResponse.pitcherType,
    throwingArm: detailsResponse.throwingArm,
    topSpeed: detailsResponse.topSpeed,
    control: detailsResponse.control,
    stamina: detailsResponse.stamina
  }
}

File: PowerUp/PowerUp.ElectronUI/Api/Shared/PlayerDetailsResponse.cs
﻿using PowerUp.Entities;
using PowerUp.Entities.Players;
using PowerUp.Generators;
using System.Text.Json.Serialization;

namespace PowerUp.ElectronUI.Api.Shared
{
  public class PlayerDetailsResponse
  {
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public EntitySourceType SourceType { get; }
    public bool CanEdit => SourceType.CanEdit();
    public int PlayerId { get; }
    public string FullName { get; }
    public string SavedName { get; }
    public IEnumerable<GeneratorWarning> GeneratedPlayer_Warnings { get; }
    public bool GeneratedPlayer_IsUnedited { get; }

    [JsonConverter(typeof(JsonStringEnumConverter))]
    public Position Position { get; }
    public string PositionAbbreviation => Position.GetAbbrev();
    public int Overall { get; }
    public string BatsAndThrows { get; }
    public string PitcherType { get; }
    public string ThrowingArm { get; }
    public string TopSpeed { get; }
    public string Control { get; }
    public string Stamina { get; }

    public PlayerDetailsResponse(Player player)
    {
      SourceType = player.SourceType;
      PlayerId = player.Id!.Value;
      FullName = player.InformalDisplayName;
      SavedName = player.SavedName;
      GeneratedPlayer_Warnings = player.GeneratorWarnings;
      GeneratedPlayer_IsUnedited = player.GeneratedPlayer_IsUnedited;
      Position = player.PrimaryPosition;
      Overall = player.Overall.RoundDown();
      BatsAndThrows = player.BatsAndThrows;
      PitcherType = player.PitcherType.GetAbbrev();
      ThrowingArm = player.ThrowingArm.GetAbbrev();
      TopSpeed = $"{player.PitcherAbilities.TopSpeedMph.RoundDown()}mph";
      Control = player.PitcherAbilities.Control.ToString();
      Stamina = player.PitcherAbilities.Stamina.ToString();
    }
  }
}
